File: CA2/ast.py
import itertools

class AST:
    def __init__(self):
        self.classes = []

    def __str__(self):
        res = str(len(self.classes)) + "\n"
        for c in self.classes:
            res += str(c)
        return res

    # Loads abstract syntax tree from iterator of lines from a .ast file,
    # Returns input iterator
    def load(self, l):
        nClasses = int(l.next())
        for i in range(nClasses):
            tClass = ASTClass()
            l = tClass.load(l)
            self.classes.append(tClass)
        return l

class ASTClass:
    def __init__(self):
        self.name = ASTIdentifier()
        self.inherit = None
        self.features = []

    def __str__(self):
        res = str(self.name)
        if self.inherit is not None:
            res += "inherits\n"
            res += str(self.inherit)
        else:
            res += "no_inherits\n"
        res += str(len(self.features)) + "\n"
        for f in self.features:
            res += str(f)
        return res

    # Loads class from iterator of lines from a .ast file,
    # Returns input iterator
    def load(self, l):
        l = self.name.load(l)
        if l.next() == "inherits":
            self.inherit = ASTIdentifier()
            l = self.inherit.load(l)
        else:
            self.inherit = None

        nFeatures = int(l.next())
        for i in range(nFeatures):
            # Need to check what the next line is without consuming it
            peek = l.next()
            l = itertools.chain([peek], l)
            feat = None
            if "attribute" in peek:
                feat = ASTAttribute()
            else:
                feat = ASTMethod()
            feat.load(l)
            self.features.append(feat)
        return l


class ASTIdentifier:
    def __init__(self):
        self.line = 0
        self.name = ""
    def __str__(self):
        return str(self.line) + "\n" + self.name + "\n"
    def load(self, l):
        self.line = int(l.next())
        self.name = l.next()
        return l

class ASTAttribute:
    def __init__(self):
        self.name = ASTIdentifier()
        self.type = ASTIdentifier()
        self.init = None
    def __str__(self):
        return ("attribute_no_init\n" if self.init is None else "attribute_init\n") + \
        str(self.name) + str(self.type) + \
        (str(self.init) if self.init is not None else "")
    def load(self,l):
        hasInit = (l.next() == "attribute_init")
        l = self.name.load(l)
        l = self.type.load(l)
        if hasInit:
            self.init = ASTExpression()
            l = self.init.load(l)
        return l

class ASTMethod:
    def __init__(self):
        self.name = ASTIdentifier()
        self.formals = [] # List of (name, type) ASTIdentifier tuples
        self.type = ASTIdentifier()
        self.body = ASTExpression()
    def __str__(self):
        res = "method\n"
        res += str(self.name)
        res += str(len(self.formals)) + "\n"
        for f in self.formals:
            res += str(f[0]) + str(f[1])
        res += str(self.type)
        res += str(self.body)
        return res
    def load(self, l):
        l.next() # Consume "method" line
        l = self.name.load(l)
        nFormals = int(l.next())
        for i in range(nFormals):
            name = ASTIdentifier()
            l = name.load(l)
            type = ASTIdentifier()
            l = type.load(l)
            self.formals.append((name,type))
        l = self.type.load(l)
        l = self.body.load(l)
        return l


class ASTExpression:
    # Classes of expressions, expressions in same class have same subparts,
    # so treated the same
    exp1 = set(['not','negate','isvoid']) # One expression subpart
    exp2 = set(['while','plus','minus','times','divide','lt','le','eq']) # Two expression subparts
    id1 =  set(['new', 'identifier']) # One identifier subpart
    def __init__(self):
        self.line = 0
        self.expr = "" # Actual expression type, as a string
        self.args = None # Expression sub-parts, single element if one subpart, tuple if multiple
    # Utility method for turning an expression list into a string
    def expListStr(self, l):
        res = str(len(l)) + "\n"
        for x in l:
            res += str(x)
        return res
    def __str__(self):
        res = str(self.line) + "\n"
        res += self.expr + "\n"
        if isinstance(self.args, tuple):
            for p in self.args:
                # p might be an expression list
                if isinstance(p, list):
                    res += self.expListStr(p)
                else:
                    res += str(p)
        else:
            # args might be an expression list
            if isinstance(self.args, list):
                res += self.expListStr(self.args)
            else:
                res += str(self.args)
        if not res[-1] == "\n":
            res += "\n"
        return res

    def loadExpList(self, l):
        nExp = int(l.next())
        mExpL = []
        for i in range(nExp):
            exp = ASTExpression()
            l = exp.load(l)
            mExpL.append(exp)
        return (mExpL, l)

    def load(self,l):
        self.line = int(l.next())
        self.expr = l.next()

        if self.expr in ASTExpression.exp1:
            self.args = ASTExpression()
            l = self.args.load(l)

        elif self.expr in ASTExpression.exp2:
            self.args = (ASTExpression(), ASTExpression())
            l = self.args[0].load(l)
            l = self.args[1].load(l)

        elif self.expr in ASTExpression.id1:
            self.args = ASTIdentifier()
            l = self.args.load(l)

        elif self.expr == "assign":
            self.args = (ASTIdentifier(), ASTExpression())
            l = self.args[0].load(l)
            l = self.args[1].load(l)

        elif self.expr == "dynamic_dispatch":
            mExp = ASTExpression()
            mId = ASTIdentifier()

            l = mExp.load(l)
            l = mId.load(l)
            mExpl, l = self.loadExpList(l)


            self.args = (mExp,mId,mExpL)

        elif self.expr == "static_dispatch":
            mExp = ASTExpression()
            mType = ASTIdentifier()
            mId = ASTIdentifier()

            l = mExp.load(l)
            l = mType.load(l)
            l = mId.load(l)
            mExpl, l = self.loadExpList(l)

            self.args = (mExp,mType,mId,mExpl)

        elif self.expr == "self_dispatch":
            mId = ASTIdentifier()

            l = mId.load(l)
            mExpl, l = self.loadExpList(l)

            self.args = (mId,mExpl)

        elif self.expr == "if":
            self.args = (ASTExpression(),ASTExpression(),ASTExpression())
            l=self.args[0].load(l)
            l=self.args[1].load(l)
            l=self.args[2].load(l)

        elif self.expr == "block":
            self.args, l = self.loadExpList(l)

        elif self.expr == "integer":
            self.args = int(l.next())

        elif self.expr == "string":
            self.args = l.next()

        # No actual arguments for true and false, but values might be useful
        elif self.expr == "true":
            self.args = True

        elif self.expr == "false":
            self.args = False

        elif self.expr == "let":
            # TODO: Something horrible
            pass

        elif self.expr == "case":
            # TODO: Also something horrible
            pass
        else:
            raise TypeError("Invalid Expression Type: " + self.expr)

        return l


import sys

if __name__ == "__main__":
    lines = sys.stdin.read()
    lines = lines.split("\n")
    ast = AST()
    ast.load(iter(lines))
    print ast

File: PA4/ast.rb
#begin annotated AST class definitions
#this implementation was artfully translated from our Python implementation for PA3
#load works assuming that .cl-ast is being loaded
#to_s works assuming that we are outputting .cl-type
class AST
    def initialize()
        @classes = []
    end

    def classes
        @classes
    end

    def to_s
        # Reference doesn't print internal classes
        return [@classes[5..-1].size,"\n", @classes[5..-1].join].join
    end

    #append built-in classes and their corresponding methods to classes list
    def preload()
        mObject = ASTClass.new.preload('Object', nil, [
            ASTMethod.new.preload('abort', [], 'Object', 'Object.abort'),
            ASTMethod.new.preload('copy', [], 'SELF_TYPE', 'Object.copy'),
            ASTMethod.new.preload('type_name', [], 'String', 'Object.type_name'),
            ])
        mIO = ASTClass.new.preload('IO', 'Object', [
            ASTMethod.new.preload('in_int', [], 'Int', 'IO.in_int'),
            ASTMethod.new.preload('in_string', [], 'String', 'IO.in_string'),
            ASTMethod.new.preload('out_int', [
                [ASTIdentifier.new.preload('x'), ASTIdentifier.new.preload('Int')]
                ], 'SELF_TYPE', 'IO.out_int'),
            ASTMethod.new.preload('out_string', [
                [ASTIdentifier.new.preload('x'), ASTIdentifier.new.preload('String')]
                ], 'SELF_TYPE', 'IO.out_string'),
            ])
        mInt = ASTClass.new.preload('Int', 'Object', [])
        mString = ASTClass.new.preload('String', 'Object', [
            ASTMethod.new.preload('concat', [
                [ASTIdentifier.new.preload('s'), ASTIdentifier.new.preload('String')]
                ], 'String', 'String.concat'),
            ASTMethod.new.preload('length', [], 'Int', 'String.length'),
            ASTMethod.new.preload('substr', [
                [ASTIdentifier.new.preload('i'), ASTIdentifier.new.preload('Int')],
                [ASTIdentifier.new.preload('l'), ASTIdentifier.new.preload('Int')]
                ], 'String', 'String.substr')
            ])
        mBool = ASTClass.new.preload('Bool', 'Object', [])
        @classes.concat([mBool, mInt, mIO, mObject, mString])
        return self
    end

    def load(en)
        nClasses = en.next.to_i
        for i in 1..nClasses
            tClass = ASTClass.new.load(en)
            @classes.push(tClass)
        end
        return self
    end
end

class ASTClass
    def initialize()
        @name = ASTIdentifier.new
        @inherit = nil
        @features = []
    end

    def name
        @name
    end

    def inherit
        @inherit
    end

    def features
        @features
    end

    def to_s
        inhstr = (@inherit.nil? ? "no_inherits\n" : ["inherits\n", @inherit].join)
        return [@name, inhstr, @features.size, "\n", @features.join].join
    end

    def preload(name, inherit, features)
        @name.preload(name)
        if !inherit.nil?
            @inherit = ASTIdentifier.new.preload(inherit)
        end
        @features = features
        return self
    end

    def load(en)
        @name.load(en)
        if en.next == 'inherits'
            @inherit = ASTIdentifier.new.load(en)
        else
            @inherit = nil
        end

        nFeatures = en.next.to_i
        for i in 1..nFeatures
            featType = en.peek
            feat = nil
            if featType.include?('attribute')
                feat = ASTBinding.new('attribute').load(en)
            else
                feat = ASTMethod.new.load(en)
            end
            @features.push(feat)
        end
        return self
    end
end

class ASTIdentifier
    def initialize()
        @line = 0
        @name = ''
    end

    def line
        @line
    end

    def name
        @name
    end

    def to_s
        return [@line,"\n",@name,"\n"].join
    end

    def preload(name)
        @name = name
        return self
    end

    def load(en)
        @line = en.next.to_i
        @name = en.next
        return self
    end
end

class ASTBinding
    def initialize(bindtype)
        @name = ASTIdentifier.new
        @type = ASTIdentifier.new
        @bindtype = bindtype    #in {'let', 'case', 'attribute'}
        @init = nil         #used in let and attribute
        @body = nil         #used in case
    end

    def name
        @name
    end

    def type
        @type
    end

    def init
        @init
    end

    def body
        @body
    end

    def to_s
        if @bindtype == 'case'
            return [@name, @type, @body].join
        elsif @bindtype == 'let'
            lethead = @init.nil? ? "let_binding_no_init\n" : "let_binding_init\n"
            return [lethead, @name, @type, @init].join
        elsif @bindtype == 'attribute'
            attrhead = @init.nil? ? "attribute_no_init\n" : "attribute_init\n"
            return [attrhead, @name, @type, @init].join
        end
    end

    def load(en)
        hasInit = (@bindtype == 'attribute' && en.next == 'attribute_init') ||
            (@bindtype == 'let' && en.next == 'let_binding_init')
        @name.load(en)
        @type.load(en)
        if @bindtype == 'case'
            @body = ASTExpression.new.load(en)
        elsif hasInit
            @init = ASTExpression.new.load(en)
        end
        return self
    end
end

class ASTMethod
    def initialize()
        @name = ASTIdentifier.new
        @formals = []
        @type = ASTIdentifier.new   #type of return value
        @body = ASTExpression.new
    end

    def name
        @name
    end

    def formals
        @formals
    end

    def type
        @type
    end

    def body
        @body
    end

    def to_s
        return ["method\n", @name, @formals.size,"\n", @formals, @type, @body].join
    end

    def preload(name, formals, type, expr)
        @name.preload(name)
        @formals = formals
        @type.preload(type)
        @body.preload([type,expr])
        return self
    end

    def load(en)
        en.next #consume "method"
        @name.load(en)
        nFormals = en.next.to_i
        for i in 1..nFormals
            fName = ASTIdentifier.new.load(en)
            fType = ASTIdentifier.new.load(en)
            @formals.push([fName, fType])
        end
        @type.load(en)
        @body.load(en)
        return self
    end
end

class ASTExpression
    @@cltype = true
    @@exp1 = ['not','negate','isvoid']
    @@exp2 = ['while','plus','minus','times','divide','lt','le','eq']
    @@id1 = ['new', 'identifier']
    def initialize()
        @line = 0
        @type = nil         #static type of the expression e.g. "Int"
        @internal = nil     #if internal expr, has string e.g. "Object.abort"
        @expr = ''          #node type e.g. "dynamic_dispatch"
        @args = nil         #data specific to the node type
    end

    def line
        @line
    end

    def type
        @type
    end

    def type=(t)
        @type = t
    end

    def expr
        @expr
    end

    def args
        @args
    end

    def to_s
        argsstr = ''
        if @args.is_a? Array
            if @args[-1].is_a? Array
                argsstr = [@args.slice(0,@args.size-1).join, @args[-1].size,"\n", @args[-1]].join
            elsif @args[0].is_a? Array
                argsstr = [@args[0].size,"\n", @args[0], @args.slice(1,@args.size).join].join
            elsif @expr == 'block'
                argsstr = [@args.size,"\n", @args].join
            else
                argsstr = @args.join
            end
        elsif @expr == 'integer' || @expr == 'string'
            argsstr = @args.to_s + "\n"
        else
            argsstr = @args.to_s
        end
        if @@cltype
            if @internal.nil?
                return [@line,"\n", @type, "\n", @expr, "\n", argsstr].join
            else
                return [@line,"\n", @type, "\n", "internal\n", @internal, "\n"].join
            end
        else
            return [@line, "\n", @expr, "\n", argsstr].join
        end
    end

    def preload(expr)
        @internal = expr[1]
        @type = expr[0]
        @expr = 'internal'
        return self
    end

    def loadExpList(en)
        nExp = en.next.to_i
        mExpL = []
        for i in 1..nExp
            exp = ASTExpression.new.load(en)
            mExpL.push(exp)
        end
        return mExpL
    end

    def load(en)
        @line = en.next.to_i
        @expr = en.next

        if @@exp1.include?(@expr)
            @args = ASTExpression.new.load(en)

        elsif @@exp2.include?(@expr)
            @args = [ASTExpression.new.load(en), ASTExpression.new.load(en)]

        elsif @@id1.include?(@expr)
            @args = ASTIdentifier.new.load(en)

        elsif @expr == 'assign'
            @args = [ASTIdentifier.new.load(en), ASTExpression.new.load(en)]

        elsif @expr == 'dynamic_dispatch'
            mExp = ASTExpression.new.load(en)
            mId = ASTIdentifier.new.load(en)
            mExpL = loadExpList(en)
            @args = [mExp, mId, mExpL]

        elsif @expr == 'static_dispatch'
            mExp = ASTExpression.new.load(en)
            mType = ASTIdentifier.new.load(en)
            mId = ASTIdentifier.new.load(en)
            mExpL = loadExpList(en)
            @args = [mExp, mType, mId, mExpL]

        elsif @expr == 'self_dispatch'
            mId = ASTIdentifier.new.load(en)
            mExpL = loadExpList(en)
            @args = [mId, mExpL]

        elsif @expr == 'if'
            @args = [ASTExpression.new.load(en),
                ASTExpression.new.load(en),
                ASTExpression.new.load(en)]

        elsif @expr == 'block'
            @args = loadExpList(en)

        elsif @expr == 'integer'
            @args = en.next.to_i

        elsif @expr == 'string'
            @args = en.next

        elsif @expr == 'true' || @expr == 'false'
            @args = ''

        elsif @expr == 'let'
            nBind = en.next.to_i
            bindList = []
            for i in 1..nBind
                bind = ASTBinding.new('let').load(en)
                bindList.push(bind)
            end
            exp = ASTExpression.new.load(en)
            @args = [bindList, exp]

        elsif @expr == 'case'
            exp = ASTExpression.new.load(en)
            nCase = en.next.to_i
            caseList = []
            for i in 1..nCase
                mCase = ASTBinding.new('case').load(en)
                caseList.push(mCase)
            end
            @args = [exp, caseList]

        else
            raise Exception.new("Invalid Expression Type #{@expr}")
        end
        return self
    end
end

#"custom" enumerator made because Ruby 1.8.5 does not have built-in support
class ASTEnumerator
    def initialize()
        @ptr = 0
    end
    def load(lines)
        @lines = lines
        return self
    end
    def next()
        line = @lines[@ptr]
        @ptr += 1
        return line
    end
    def peek()
        return @lines[@ptr]
    end
end

if __FILE__ == $0
    lines = File.read(ARGV[0]).split("\n")
    ast = AST.new.preload.load(lines.each)
    #File.write(ARGV[0]+"-echo", "#{ast}")  #too new for legacy server
    tmap = TypeMaps.new.load(ast)
    if tmap.err.nil?
        adhocErr = checkAdhoc(tmap)
        if adhocErr.nil?
            puts 'class map'
            tmap.cmap.each do |k,v|
                puts k
                v.each do |vv|
                    puts vv.name.name
                end
            end
            puts 'implementation map'
            tmap.imap.each do |k,v|
                puts k
                v.each do |vname, vfeat|
                    puts vname+" "+vfeat.name.name
                end
            end
            puts 'parent map'
            tmap.pmap.each do |k,v|
                puts k.name.name
                puts v.name.name
            end
        else
            puts adhocErr
        end
    else
        puts tmap.err
    end
end
